modules/sdk/src/main/scala/org/tessellation/sdk/infrastructure/snapshot/CurrencySnapshotCreator.scala

```diff
@@ -50,137 +54,181 @@ trait CurrencySnapshotCreator[F[_]] {
 
-        (height, subHeight) <- getHeightAndSubHeight(lastArtifact, deprecated, remainedActive, accepted)
-
-        artifact = CurrencyIncrementalSnapshot(
-          currentOrdinal,
-          height,
-          subHeight,
-          lastArtifactHash,
-          accepted,
-          acceptedRewardTxs,
-          SnapshotTips(
-            deprecated = deprecated,
-            remainedActive = remainedActive
-          ),
-          stateProof,
-          currentEpochProgress,
-          maybeNewDataState
-        )
-      } yield
-        CurrencySnapshotCreationResult(
-          artifact,
-          CurrencySnapshotContext(lastContext.address, snapshotInfo),
-          awaitingBlocks.toSet,
-          rejectedBlocks.toSet
-        )
+          maybeLastDataState <- (maybeLastDataApplication, maybeDataApplicationService).mapN {
+            case (lastDataApplication, (_, service)) =>
+              service
+                .deserializeState(lastDataApplication)
+                .flatTap {
+                  case Left(err) => logger.warn(err)(s"Cannot deserialize custom state")
+                  case Right(a)  => logger.info(s"Deserialized state: $a")
+                }
+                .map(_.toOption)
+                .handleErrorWith(err =>
+                  logger.error(err)(s"Unhandled exception during deserialization data application, fallback to empty state") >>
+                    none[DataState].pure[F]
+                )
+          }.flatSequence
+
+          maybeNewDataState <- (maybeDataApplicationService, maybeLastDataState).mapN {
+            case (((nodeContext, service), lastState)) =>
+              implicit val context: L0NodeContext[F] = nodeContext
+              NonEmptyList
+                .fromList(dataBlocks.distinctBy(_.value.roundId))
+                .map(_.flatMap(_.value.updates))
+                .map { updates =>
+                  service
+                    .validateData(lastState, updates)
+                    .flatTap { validated =>
+                      Applicative[F].unlessA(validated.isValid)(logger.warn(s"Data application is invalid, errors: ${validated.toString}"))
+                    }
+                    .map(_.isValid)
+                    .handleErrorWith(err =>
+                      logger.error(err)(s"Unhandled exception during validating data application, assumed as invalid") >> false.pure[F]
+                    )
+                    .ifM(
+                      service
+                        .combine(lastState, updates.toList)
+                        .flatMap { state =>
+                          service.serializeState(state)
+                        }
+                        .map(_.some)
+                        .handleErrorWith(err =>
+                          logger.error(err)(
+                            s"Unhandled exception during combine and serialize data application, fallback to last data application"
+                          ) >> maybeLastDataApplication.pure[F]
```

Comment:
Same as above. Use `.as` and wonder if this should be WARN

Reply:
@marcinwadon ^ same question regarding severity